Adding BlocklyDevTools.Analytics

AnmAtAnm opened the pull request:
Adding BlocklyDevTools.Analytics, an interface for integrating an analytics library to track basic usage, including:
 * navigation.
 * saving, importing, and exporting.
 * warnings and errors.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly/1217)
<!-- Reviewable:end -->

rachel-fenichel commented on demos/blockfactory/analytics.js:
```diff
@@ -0,0 +1,102 @@
+ */
+
+/**
+ * @fileoverview Stubbed interface functions for analytics integration.
```
Missing @author

rachel-fenichel commented on demos/blockfactory/analytics.js:
```diff
@@ -0,0 +1,102 @@
+BlocklyDevTools.Analytics.LOG_TO_CONSOLE = false;
+
+/**
+ * Initializes the analytics framework, including noting that the page/app was
```
Missing @private/@package/@public for everything.

rachel-fenichel commented on demos/blockfactory/analytics.js:
```diff
@@ -0,0 +1,102 @@
+/**
+ * Event noting a project resource was saved. In the web Block Factory, this
+ * means downloaded to the user's system.
+ * @param typeId {string} An identifying string for the exported type.
```
Missing doc for optFormalId param.

rachel-fenichel commented on demos/blockfactory/app_controller.js:
```diff
@@ -137,9 +140,12 @@ AppController.prototype.exportBlockLibraryToFile = function() {
   // Download file if all necessary parameters are provided.
   if (filename) {
     FactoryUtils.createAndDownloadFile(blockLibText, filename, 'xml');
+    BlocklyDevTools.Analytics.onExport('BlockFactoryLibrary.xml');
```
The strings that indicate the type should be constants defined in BlocklyDevTools.Analytics.  That can be done in a followup as you flesh this out.

AnmAtAnm commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[demos/blockfactory/analytics.js, line 22 at r1](https://reviewable.io:443/reviews/google/blockly/1217#-Koic8LXJLO19FEVbUl7-r1-22:-Komhnb8_C54st6DzR_0:b-gfm10b) ([raw file](https://github.com/google/blockly/blob/c15aee18b53e254934ab9f8ada2fe6daa2a8b44e/demos/blockfactory/analytics.js#L22)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Missing @author
</blockquote></details>

Avoiding duplication and any eventual inaccuracy.  One can always look to git blame.

---

*[demos/blockfactory/analytics.js, line 30 at r1](https://reviewable.io:443/reviews/google/blockly/1217#-Koic8LXJLO19FEVbUl7-r1-30:-KomhoyzaQQlV8rScheG:b-896fix) ([raw file](https://github.com/google/blockly/blob/c15aee18b53e254934ab9f8ada2fe6daa2a8b44e/demos/blockfactory/analytics.js#L30)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Missing @private/@package/@public for everything.
</blockquote></details>

Done.

---

*[demos/blockfactory/analytics.js, line 69 at r1](https://reviewable.io:443/reviews/google/blockly/1217#-Koic8LXJLO19FEVbUl7-r1-69:-Komi3Bj6Nsmaquhm9q3:b-k1w1zb) ([raw file](https://github.com/google/blockly/blob/c15aee18b53e254934ab9f8ada2fe6daa2a8b44e/demos/blockfactory/analytics.js#L69)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Missing doc for optFormalId param.
</blockquote></details>

Hmm... Forgot I even added this.  Replaced it with optMetadata to also store platform info.
Added JSDoc.

---

*[demos/blockfactory/app_controller.js, line 143 at r1](https://reviewable.io:443/reviews/google/blockly/1217#-Koic8LXJLO19FEVbUl8-r1-143:-KompOEK74otVSfUwNhu:b-896fix) ([raw file](https://github.com/google/blockly/blob/c15aee18b53e254934ab9f8ada2fe6daa2a8b44e/demos/blockfactory/app_controller.js#L143)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

The strings that indicate the type should be constants defined in BlocklyDevTools.Analytics.  That can be done in a followup as you flesh this out.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1217)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Missing @const {string} on all of your constants.

I expected your functions to be package, not public.

---

Reviewed 2 of 6 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1217#-:-KomvTeOeGUQIBs-XMpx:biwvy5l)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
They need to be accessible from `workspacefactory` directory, at least for `wfactory_controller`.

---

Review status: 5 of 6 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1217#-:-KonSAgFZ6ghzDVKSuEE:b-451q8m)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1217#-:-Konmxmv31OFkVYrm9bJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->
